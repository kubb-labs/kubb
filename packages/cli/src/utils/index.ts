export * from './getConfig.ts'
export * from './getPlugins.ts'
export * from './watcher.ts'
export * from './getCosmiConfig.ts'
export * from './parseHrtimeToSeconds.ts'
export * from './parseText.ts'
