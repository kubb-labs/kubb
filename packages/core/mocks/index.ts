import { pascalCase } from '../src/transformers/casing.ts'

import type { PluginManager } from '../src/PluginManager.ts'
import { readSync } from '../src/fs/read.ts'

export const mockedPluginManager = {
  resolveName: ({ name, type }) => {
    if (type === 'type') {
      return pascalCase(name)
    }

    return name
  },
  resolvePath: ({ baseName }) => baseName,
  logger: {
    emit(message) {
      console.log(message)
    },
    on(eventName, args) {},
    logLevel: 'info',
  },
  getFile: ({ name, extName, pluginKey }) => {
    const baseName = `${name}${extName}`
    let source = ''

    try {
      source = readSync(baseName)
    } catch (_e) {
      //
    }

    return {
      path: baseName,
      baseName,
      meta: {
        pluginKey,
      },
      source,
    }
  },
} as PluginManager
