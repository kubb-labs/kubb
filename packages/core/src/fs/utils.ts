import { normalize, relative } from 'node:path'

function slash(path: string, platform: 'windows' | 'mac' | 'linux' = 'linux') {
  const isWindowsPath = /^\\\\\?\\/.test(path)
  const normalizedPath = normalize(path)

  if (['linux', 'mac'].includes(platform) && !isWindowsPath) {
    // linux and mac
    return normalizedPath.replaceAll(/\\/g, '/').replace('../', '')
  }

  // windows
  return normalizedPath.replaceAll(/\\/g, '/').replace('../', '')
}

export function getRelativePath(rootDir?: string | null, filePath?: string | null, platform: 'windows' | 'mac' | 'linux' = 'linux'): string {
  if (!rootDir || !filePath) {
    throw new Error(`Root and file should be filled in when retrieving the relativePath, ${rootDir || ''} ${filePath || ''}`)
  }

  const relativePath = relative(rootDir, filePath)

  // On Windows, paths are separated with a "\"
  // However, web browsers use "/" no matter the platform
  const slashedPath = slash(relativePath, platform)

  if (slashedPath.startsWith('../')) {
    return slashedPath
  }

  return `./${slashedPath}`
}
