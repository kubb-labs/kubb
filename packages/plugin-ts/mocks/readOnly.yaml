openapi: 3.0.0

components:
  schemas:
    Demo:
      type: object
      required:
        - attribute_readonly_required
        - attribute_required
        - attribute
        - attribute_with_ref_readonly_required
      additionalProperties: false
      properties:
        attribute_readonly:
          type: integer
          readOnly: true
        attribute_writeOnly:
          description: not required
          type: integer
          writeOnly:  true
        attribute_with_ref:
          $ref: '#/components/schemas/attribute_readonly'
        attribute_with_ref_readonly:
          $ref: '#/components/schemas/attribute_writeOnly'
    attribute_readonly:
      type: integer
      readOnly: true
    attribute_writeOnly:
      description: not required
      type: integer
      writeOnly: true
