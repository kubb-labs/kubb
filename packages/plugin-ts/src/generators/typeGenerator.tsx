import { createReactGenerator, type OperationSchema as OperationSchemaType, type OperationSchemas, SchemaGenerator, schemaKeywords } from '@kubb/plugin-oas'
import { Type } from '../components'
import type { PluginTs } from '../types'
import { File, useApp } from '@kubb/react'
import { useOas, useOperationManager, useSchemaManager } from '@kubb/plugin-oas/hooks'
import { pluginTsName } from '@kubb/plugin-ts'
import { Oas } from '@kubb/plugin-oas/components'
import type { PluginManager } from '@kubb/core'
import { print, type ts } from '@kubb/parser-ts'
import * as factory from '@kubb/parser-ts/factory'
import transformers from '@kubb/core/transformers'

function printCombinedSchema({ name, schemas, pluginManager }: { name: string; schemas: OperationSchemas; pluginManager: PluginManager }): string {
  const properties: Record<string, ts.TypeNode> = {}

  if (schemas.response) {
    const identifier = pluginManager.resolveName({
      name: schemas.response.name,
      pluginKey: [pluginTsName],
      type: 'function',
    })
    properties['response'] = factory.createTypeReferenceNode(factory.createIdentifier(identifier), undefined)
  }

  if (schemas.request) {
    const identifier = pluginManager.resolveName({
      name: schemas.request.name,
      pluginKey: [pluginTsName],
      type: 'function',
    })
    properties['request'] = factory.createTypeReferenceNode(factory.createIdentifier(identifier), undefined)
  }

  if (schemas.pathParams) {
    const identifier = pluginManager.resolveName({
      name: schemas.pathParams.name,
      pluginKey: [pluginTsName],
      type: 'function',
    })
    properties['pathParams'] = factory.createTypeReferenceNode(factory.createIdentifier(identifier), undefined)
  }

  if (schemas.queryParams) {
    const identifier = pluginManager.resolveName({
      name: schemas.queryParams.name,
      pluginKey: [pluginTsName],
      type: 'function',
    })
    properties['queryParams'] = factory.createTypeReferenceNode(factory.createIdentifier(identifier), undefined)
  }

  if (schemas.headerParams) {
    const identifier = pluginManager.resolveName({
      name: schemas.headerParams.name,
      pluginKey: [pluginTsName],
      type: 'function',
    })
    properties['headerParams'] = factory.createTypeReferenceNode(factory.createIdentifier(identifier), undefined)
  }

  if (schemas.errors) {
    properties['errors'] = factory.createUnionDeclaration({
      nodes: schemas.errors.map((error) => {
        const identifier = pluginManager.resolveName({
          name: error.name,
          pluginKey: [pluginTsName],
          type: 'function',
        })

        return factory.createTypeReferenceNode(factory.createIdentifier(identifier), undefined)
      }),
    })!
  }

  const namespaceNode = factory.createTypeAliasDeclaration({
    name,
    type: factory.createTypeLiteralNode(
      Object.keys(properties)
        .map((key) => {
          const type = properties[key]
          if (!type) {
            return undefined
          }

          return factory.createPropertySignature({
            name: transformers.pascalCase(key),
            type,
          })
        })
        .filter(Boolean),
    ),
    modifiers: [factory.modifiers.export],
  })

  return print([namespaceNode])
}

export const typeGenerator = createReactGenerator<PluginTs>({
  name: 'typescript',
  Operation({ operation, options }) {
    const { mapper, enumType, optionalType } = options

    const { plugin, pluginManager, mode } = useApp<PluginTs>()
    const oas = useOas()
    const { getSchemas, getFile, getName } = useOperationManager()
    const schemaManager = useSchemaManager()

    const file = getFile(operation)
    const schemas = getSchemas(operation)
    const type = getName(operation, { type: 'function', pluginKey: [pluginTsName] })
    const combinedSchemaName = operation.method === 'get' ? `${type}Query` : `${type}Mutation`
    const schemaGenerator = new SchemaGenerator(options, {
      oas,
      plugin,
      pluginManager,
      mode,
      override: options.override,
    })

    const operationSchemas = [schemas.pathParams, schemas.queryParams, schemas.headerParams, schemas.statusCodes, schemas.request, schemas.response]
      .flat()
      .filter(Boolean)

    const mapOperationSchema = ({ name, schema, description, keysToOmit, ...options }: OperationSchemaType, i: number) => {
      // hack so Params can be optional when needed
      const tree = schemaGenerator.parse({ schema, name })
      const imports = schemaManager.getImports(tree)

      const type = {
        name: schemaManager.getName(name, { type: 'type' }),
        typedName: schemaManager.getName(name, { type: 'type' }),
        file: schemaManager.getFile(options.operationName || name, { tag: options.operation?.getTags()[0]?.name }),
      }

      return (
        <Oas.Schema key={i} name={name} value={schema} tree={tree}>
          {mode === 'split' && imports.map((imp, index) => <File.Import key={index} root={file.path} path={imp.path} name={imp.name} isTypeOnly />)}
          <Type
            name={type.name}
            typedName={type.typedName}
            description={description}
            tree={tree}
            mapper={mapper}
            enumType={enumType}
            optionalType={optionalType}
            keysToOmit={keysToOmit}
          />
        </Oas.Schema>
      )
    }

    return (
      <File baseName={file.baseName} path={file.path} meta={file.meta}>
        {operationSchemas.map(mapOperationSchema)}

        <File.Source name={combinedSchemaName} isExportable isIndexable isTypeOnly>
          {printCombinedSchema({ name: combinedSchemaName, schemas, pluginManager })}
        </File.Source>
      </File>
    )
  },
  Schema({ schema, options }) {
    const { mapper, enumType, optionalType } = options
    const { mode } = useApp<PluginTs>()

    const { getName, getImports, getFile } = useSchemaManager()
    const imports = getImports(schema.tree)

    if (enumType === 'asPascalConst') {
      console.warn(`enumType '${enumType}' is deprecated`)
    }

    const type = {
      name: getName(schema.name, { type: 'function' }),
      typedName: getName(schema.name, { type: 'type' }),
      file: getFile(schema.name),
    }

    return (
      <File baseName={type.file.baseName} path={type.file.path} meta={type.file.meta}>
        {mode === 'split' && imports.map((imp, index) => <File.Import key={index} root={type.file.path} path={imp.path} name={imp.name} isTypeOnly />)}
        <Type
          name={type.name}
          typedName={type.typedName}
          description={schema.value.description}
          tree={schema.tree}
          mapper={mapper}
          enumType={enumType}
          optionalType={optionalType}
        />
      </File>
    )
  },
})
