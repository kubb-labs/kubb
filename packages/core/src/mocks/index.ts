import path from 'node:path'
import { getSource } from '../FileManager'
import type { File, ResolvedFile } from '../fs/types.ts'
import type { Logger } from '../logger'
import type { PluginManager } from '../PluginManager.ts'
import { camelCase, pascalCase } from '../transformers/casing.ts'
import type { Plugin } from '../types.ts'
import { createFile } from '../utils'

export const mockedLogger = {
  emit(_type, _message) {},
  on(_type, _message) {},
  consola: {},
} as Logger

export const createMockedPluginManager = (name?: string) =>
  ({
    resolveName: (result) => {
      if (result.type === 'file') {
        return camelCase(name || result.name)
      }

      if (result.type === 'type') {
        return pascalCase(result.name)
      }

      if (result.type === 'function') {
        return camelCase(result.name)
      }

      return camelCase(result.name)
    },
    config: {
      output: {
        path: './path',
      },
    },
    resolvePath: ({ baseName }) => baseName,
    logger: {
      emit(message) {
        console.log(message)
      },
      on(_eventName, _args) {},
      logLevel: 3,
    },
    getPluginByKey: (_pluginKey: Plugin['key']) => {
      return undefined
    },
    getFile: ({ name, extname, pluginKey }) => {
      const baseName = `${name}${extname}`

      return {
        path: baseName,
        baseName,
        meta: {
          pluginKey,
        },
      }
    },
  }) as PluginManager

export const mockedPluginManager = createMockedPluginManager('')

export async function matchFiles(files: Array<ResolvedFile | File> | undefined, pre?: string) {
  if (!files) {
    return undefined
  }

  for await (const file of files) {
    const source = await getSource(createFile(file), { logger: mockedLogger })
    await expect(source).toMatchFileSnapshot(path.join(...['__snapshots__', pre, file.path].filter(Boolean)))
  }
}
