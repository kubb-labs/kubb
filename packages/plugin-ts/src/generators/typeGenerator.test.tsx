import { createMockedPluginManager, matchFiles, mockedPluginManager } from '@kubb/core/mocks'

import path from 'node:path'
import type { Plugin } from '@kubb/core'
import type { HttpMethod } from '@kubb/oas'
import { parse } from '@kubb/oas/parser'
import { OperationGenerator, SchemaGenerator } from '@kubb/plugin-oas'
import { getSchemas } from '@kubb/plugin-oas/utils'
import ts, { factory } from 'typescript'
import type { PluginTs } from '../types.ts'
import { typeGenerator } from './typeGenerator.tsx'

describe('typeGenerator schema', async () => {
  const testData = [
    {
      name: 'PetQuestionToken',
      input: '../../mocks/petStore.yaml',
      path: 'Pet',
      options: {
        optionalType: 'questionToken',
      },
    },
    {
      name: 'PetUndefined',
      input: '../../mocks/petStore.yaml',
      path: 'Pet',
      options: {
        optionalType: 'undefined',
      },
    },
    {
      name: 'PetMapper',
      input: '../../mocks/petStore.yaml',
      path: 'Pet',
      options: {
        mapper: {
          category: factory.createPropertySignature(
            undefined,
            factory.createIdentifier('category'),
            factory.createToken(ts.SyntaxKind.QuestionToken),
            factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
          ),
        },
      },
    },
    {
      name: 'PetQuestionTokenAndUndefined',
      input: '../../mocks/petStore.yaml',
      path: 'Pet',
      options: {
        optionalType: 'questionTokenAndUndefined',
      },
    },
    {
      name: 'Pets',
      input: '../../mocks/petStore.yaml',
      path: 'Pets',
      options: {
        optionalType: 'questionToken',
      },
    },
    {
      name: 'PetsStoreRef',
      input: '../../mocks/petStoreRef.yaml',
      path: 'Pets',
      options: {
        optionalType: 'questionToken',
      },
    },
    {
      name: 'PetsStoreDiscriminator',
      input: '../../mocks/discriminator.yaml',
      path: 'Petstore',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'CatTypeAsConst',
      input: '../../mocks/discriminator.yaml',
      path: 'Cat',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'DogTypeAsConst',
      input: '../../mocks/discriminator.yaml',
      path: 'Dog',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'NullConstNull',
      input: '../../mocks/discriminator.yaml',
      path: 'NullConst',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'StringValueConst',
      input: '../../mocks/discriminator.yaml',
      path: 'StringValueConst',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'NumberValueConst',
      input: '../../mocks/discriminator.yaml',
      path: 'NumberValueConst',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'MixedValueTypeConst',
      input: '../../mocks/discriminator.yaml',
      path: 'MixedValueTypeConst',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumVarNamesType',
      input: '../../mocks/enums.yaml',
      path: 'enumVarNames.Type',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumNamesType',
      input: '../../mocks/enums.yaml',
      path: 'enumNames.Type',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumItems',
      input: '../../mocks/enums.yaml',
      path: 'enum.Items',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumString',
      input: '../../mocks/enums.yaml',
      path: 'enum.String',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumNullableMember',
      input: '../../mocks/enums.yaml',
      path: 'enum.NullableMember',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumNullableType',
      input: '../../mocks/enums.yaml',
      path: 'enum.NullableType',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumAllOf',
      input: '../../mocks/enums.yaml',
      path: 'enum.AllOf',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumInObject',
      input: '../../mocks/enums.yaml',
      path: 'enum.InObject',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumArray',
      input: '../../mocks/enums_2.0.yaml',
      path: 'enum.Array',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumNames',
      input: '../../mocks/enums.yaml',
      path: 'enumNames.Type',
      options: {
        enumType: 'enum',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumNamesPascalConst',
      input: '../../mocks/enums.yaml',
      path: 'enumNames.Type',
      options: {
        enumType: 'asPascalConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumNamesConst',
      input: '../../mocks/enums.yaml',
      path: 'enumNames.Type',
      options: {
        enumType: 'constEnum',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'EnumNamesLiteral',
      input: '../../mocks/enums.yaml',
      path: 'enumNames.Type',
      options: {
        enumType: 'literal',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'Body_upload_file_api_assets_post',
      input: '../../mocks/typeAssertions.yaml',
      path: 'Body_upload_file_api_assets_post',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'Plain_file',
      input: '../../mocks/typeAssertions.yaml',
      path: 'Plain_file',
      options: {
        enumType: 'asConst',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'Plain_dateString',
      input: '../../mocks/typeAssertions.yaml',
      path: 'Plain_date',
      options: {
        enumType: 'asConst',
        dateType: 'string',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'Plain_dateDate',
      input: '../../mocks/typeAssertions.yaml',
      path: 'Plain_date',
      options: {
        enumType: 'asConst',
        dateType: 'date',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'Plain_timeDate',
      input: '../../mocks/typeAssertions.yaml',
      path: 'Plain_time',
      options: {
        enumType: 'asConst',
        dateType: 'date',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'Plain_email',
      input: '../../mocks/typeAssertions.yaml',
      path: 'Plain_email',
      options: {
        enumType: 'asConst',
        dateType: 'date',
        optionalType: 'questionToken',
      },
    },
    {
      name: 'Plain_uuid',
      input: '../../mocks/typeAssertions.yaml',
      path: 'Plain_uuid',
      options: {
        enumType: 'asConst',
        dateType: 'date',
        optionalType: 'questionToken',
      },
    },
  ] as const satisfies Array<{
    input: string
    name: string
    path: string
    options: Partial<PluginTs['resolvedOptions']>
  }>

  test.each(testData)('$name', async (props) => {
    const oas = await parse(path.resolve(__dirname, props.input))

    const options: PluginTs['resolvedOptions'] = {
      usedEnumNames: {},
      enumType: 'asConst',
      enumSuffix: 'enum',
      dateType: 'string',
      transformers: {},
      oasType: false,
      unknownType: 'any',
      optionalType: 'questionToken',
      override: [],
      mapper: {},
      ...props.options,
    }
    const plugin = { options } as Plugin<PluginTs>
    const instance = new SchemaGenerator(options, {
      oas,
      pluginManager: createMockedPluginManager(props.name),
      plugin,
      contentType: 'application/json',
      include: undefined,
      override: undefined,
      mode: 'split',
      output: './gen',
    })
    await instance.build(typeGenerator)

    const schemas = getSchemas({ oas })
    const name = props.path
    const schema = schemas[name]!
    const tree = instance.parse({ schema, name })

    const files = await typeGenerator.schema?.({
      schema: {
        name,
        tree,
        value: schema,
      },
      options,
      instance,
    })

    await matchFiles(files)
  })
})

describe('typeGenerator operation', async () => {
  const testData = [
    {
      name: 'showPetById',
      input: '../../mocks/petStore.yaml',
      path: '/pets/{petId}',
      method: 'get',
      options: {},
    },
    {
      name: 'getPets',
      input: '../../mocks/petStore.yaml',
      path: '/pets',
      method: 'get',
      options: {},
    },
    {
      name: 'createPet',
      input: '../../mocks/petStore.yaml',
      path: '/pets',
      method: 'post',
      options: {},
    },
    {
      name: 'createPet with unknownType unknown',
      input: '../../mocks/petStore.yaml',
      path: '/pets',
      method: 'post',
      options: {
        unknownType: 'unknown',
      },
    },
    {
      name: 'deletePet',
      input: '../../mocks/petStore.yaml',
      path: '/pets/{petId}',
      method: 'delete',
      options: {},
    },
  ] as const satisfies Array<{
    input: string
    name: string
    path: string
    method: HttpMethod
    options: Partial<PluginTs['resolvedOptions']>
  }>

  test.each(testData)('$name', async (props) => {
    const oas = await parse(path.resolve(__dirname, props.input))

    const options: PluginTs['resolvedOptions'] = {
      enumType: 'asConst',
      enumSuffix: '',
      dateType: 'string',
      optionalType: 'questionToken',
      usedEnumNames: {},
      transformers: {},
      oasType: false,
      unknownType: 'any',
      override: [],
      mapper: {},
      ...props.options,
    }
    const plugin = { options } as Plugin<PluginTs>
    const instance = new OperationGenerator(options, {
      oas,
      include: undefined,
      pluginManager: createMockedPluginManager(props.name),
      plugin,
      contentType: undefined,
      override: undefined,
      mode: 'split',
      exclude: [],
    })
    const operation = oas.operation(props.path, props.method)
    const files = await typeGenerator.operation?.({
      operation,
      options,
      instance,
    })

    await matchFiles(files)
  })
})
