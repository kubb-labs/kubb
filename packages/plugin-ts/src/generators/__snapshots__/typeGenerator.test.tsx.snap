// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`typeGenerator operation > 'createPet with unknownType unknown' 1`] = `
"/**
 * @description Null response
*/
export type CreatePets201 = (error & {
    name?: errorCode;
});

/**
 * @description unexpected error
*/
export type CreatePetsError = error;

export type CreatePetsMutationRequest = {
    /**
     * @type string
    */
    name: string;
    /**
     * @type string
    */
    tag: string;
};

/**
 * @description Null response
*/
export type CreatePetsMutationResponse = (error & {
    name?: errorCode;
});

export type createPetsMutation = {
    Response: createPetsMutationResponse;
    Request: createPetsMutationRequest;
};"
`;

exports[`typeGenerator operation > 'createPet' 1`] = `
"/**
 * @description Null response
*/
export type CreatePets201 = (error & {
    name?: errorCode;
});

/**
 * @description unexpected error
*/
export type CreatePetsError = error;

export type CreatePetsMutationRequest = {
    /**
     * @type string
    */
    name: string;
    /**
     * @type string
    */
    tag: string;
};

/**
 * @description Null response
*/
export type CreatePetsMutationResponse = (error & {
    name?: errorCode;
});

export type createPetsMutation = {
    Response: createPetsMutationResponse;
    Request: createPetsMutationRequest;
};"
`;

exports[`typeGenerator operation > 'deletePet' 1`] = `
"export type DeletePetsPetidMutationResponse = any;

export type deletePetsPetidMutation = {
    Response: deletePetsPetidMutationResponse;
};"
`;

exports[`typeGenerator operation > 'getPets' 1`] = `
"export type ListPetsQueryParams = {
    /**
     * @description How many items to return at one time (max 100)
     * @type string | undefined
    */
    limit?: string;
};

/**
 * @description A paged array of pets
*/
export type ListPets200 = pets;

/**
 * @description unexpected error
*/
export type ListPetsError = error;

/**
 * @description A paged array of pets
*/
export type ListPetsQueryResponse = pets;

export type listPetsQuery = {
    Response: listPetsQueryResponse;
    QueryParams: listPetsQueryParams;
};"
`;

exports[`typeGenerator operation > 'showPetById' 1`] = `
"export type ShowPetByIdPathParams = {
    /**
     * @description The id of the pet to retrieve
     * @type string
    */
    petId: string;
    /**
     * @description The id of the pet to retrieve
     * @type string
    */
    testId: string;
};

/**
 * @description Expected response to a valid request
*/
export type ShowPetById200 = pet;

/**
 * @description unexpected error
*/
export type ShowPetByIdError = error;

/**
 * @description Expected response to a valid request
*/
export type ShowPetByIdQueryResponse = pet;

export type showPetByIdQuery = {
    Response: showPetByIdQueryResponse;
    PathParams: showPetByIdPathParams;
};"
`;

exports[`typeGenerator schema > 'Body_upload_file_api_assets_post' 1`] = `
"export type bodyUploadFileApiAssetsPost = {
    /**
     * @type string, binary
    */
    file: Blob;
};"
`;

exports[`typeGenerator schema > 'CatTypeAsConst' 1`] = `
"export type cat = {
    /**
     * @type string, uuid
    */
    id: string;
    type: "Cat";
    /**
     * @type boolean | undefined
    */
    meow?: boolean;
};"
`;

exports[`typeGenerator schema > 'DogTypeAsConst' 1`] = `
"export type dog = {
    /**
     * @type string, uuid
    */
    id: string;
    /**
     * @type number | undefined
    */
    bark?: number;
    type: "Dog";
};"
`;

exports[`typeGenerator schema > 'EnumAllOf' 1`] = `
"export const enumAllOfReasonEnum2 = {
    "created_at": "created_at",
    "description": "description"
} as const;

export type EnumAllOfReasonEnum2 = (typeof enumAllOfReasonEnum2)[keyof typeof enumAllOfReasonEnum2];

export type enumAllOf = {
    reason?: EnumAllOfReasonEnum2;
};"
`;

exports[`typeGenerator schema > 'EnumArray' 1`] = `
"export const enumArrayIdentifierEnum2 = {
    "NW": "NW",
    "NE": "NE",
    "SW": "SW",
    "SE": "SE"
} as const;

export type EnumArrayIdentifierEnum2 = (typeof enumArrayIdentifierEnum2)[keyof typeof enumArrayIdentifierEnum2];

export type enumArray = {
    /**
     * @type array | undefined
    */
    identifier?: [
        number,
        string,
        EnumArrayIdentifierEnum2
    ];
};"
`;

exports[`typeGenerator schema > 'EnumInObject' 1`] = `
"export const enumInObjectReasonEnum2 = {
    "created_at": "created_at",
    "description": "description"
} as const;

export type EnumInObjectReasonEnum2 = (typeof enumInObjectReasonEnum2)[keyof typeof enumInObjectReasonEnum2];

export type enumInObject = {
    /**
     * @type string | undefined
    */
    reason?: EnumInObjectReasonEnum2;
};"
`;

exports[`typeGenerator schema > 'EnumItems' 1`] = `
"export const enumItemsEnum2 = {
    "created_at": "created_at",
    "description": "description"
} as const;

export type EnumItemsEnum2 = (typeof enumItemsEnum2)[keyof typeof enumItemsEnum2];

export type enumItems = EnumItemsEnum2[];"
`;

exports[`typeGenerator schema > 'EnumNames' 1`] = `
"export enum EnumNamesTypeEnum2 {
    "Pending" = 0,
    "Received" = 1
}

export type enumNamesType = EnumNamesTypeEnum2;"
`;

exports[`typeGenerator schema > 'EnumNamesConst' 1`] = `
"export const enum EnumNamesTypeEnum2 {
    "Pending" = 0,
    "Received" = 1
}

export type enumNamesType = EnumNamesTypeEnum2;"
`;

exports[`typeGenerator schema > 'EnumNamesLiteral' 1`] = `
"export type EnumNamesTypeEnum2 = 0 | 1;

export type enumNamesType = EnumNamesTypeEnum2;"
`;

exports[`typeGenerator schema > 'EnumNamesPascalConst' 1`] = `
"export const EnumNamesTypeEnum2 = {
    "Pending": 0,
    "Received": 1
} as const;

type EnumNamesTypeEnum2 = (typeof EnumNamesTypeEnum2)[keyof typeof EnumNamesTypeEnum2];

export type enumNamesType = EnumNamesTypeEnum2;"
`;

exports[`typeGenerator schema > 'EnumNamesType' 1`] = `
"export const enumNamesType = {
    "Pending": 0,
    "Received": 1
} as const;

export type EnumNamesTypeEnum2 = (typeof enumNamesType)[keyof typeof enumNamesType];

export type enumNamesType = EnumNamesTypeEnum2;"
`;

exports[`typeGenerator schema > 'EnumNullableMember' 1`] = `
"export const enumNullableMemberEnum2 = {
    "first": "first",
    "second": "second"
} as const;

export type EnumNullableMemberEnum2 = (typeof enumNullableMemberEnum2)[keyof typeof enumNullableMemberEnum2];

export type enumNullableMember = EnumNullableMemberEnum2 | null;"
`;

exports[`typeGenerator schema > 'EnumNullableType' 1`] = `
"export const enumNullableTypeEnum2 = {
    "first": "first",
    "second": "second"
} as const;

export type EnumNullableTypeEnum2 = (typeof enumNullableTypeEnum2)[keyof typeof enumNullableTypeEnum2];

export type enumNullableType = EnumNullableTypeEnum2 | null;"
`;

exports[`typeGenerator schema > 'EnumString' 1`] = `
"export const enumStringEnum2 = {
    "created_at": "created_at",
    "description": "description",
    "FILE.UPLOADED": "FILE.UPLOADED",
    "FILE.DOWNLOADED": "FILE.DOWNLOADED"
} as const;

export type EnumStringEnum2 = (typeof enumStringEnum2)[keyof typeof enumStringEnum2];

export type enumString = EnumStringEnum2;"
`;

exports[`typeGenerator schema > 'EnumVarNamesType' 1`] = `
"export const enumVarNamesTypeEnum2 = {
    "Pending": 0,
    "Received": 1
} as const;

export type EnumVarNamesTypeEnum2 = (typeof enumVarNamesTypeEnum2)[keyof typeof enumVarNamesTypeEnum2];

export type enumVarNamesType = EnumVarNamesTypeEnum2;"
`;

exports[`typeGenerator schema > 'MixedValueTypeConst' 1`] = `
"/**
 * @description This probably should fail miserably
*/
export type mixedValueTypeConst = {
    /**
     * @type number | undefined
    */
    foobar?: "foobar";
};"
`;

exports[`typeGenerator schema > 'NullConstNull' 1`] = `
"/**
 * @description The value of this keyword MAY be null. \\nUse of this keyword is functionally equivalent to an \\"enum\\" (Section 6.1.2) with a single value.\\nAn instance validates successfully against this keyword if its value is equal to the value of the keyword.\\n
*/
export type nullConst = {
    withoutValue?: null;
};"
`;

exports[`typeGenerator schema > 'NumberValueConst' 1`] = `
"/**
 * @description its value is equal to the value of the keyword
*/
export type numberValueConst = {
    /**
     * @type number | undefined
    */
    foobar?: 42;
};"
`;

exports[`typeGenerator schema > 'PetQuestionToken' 1`] = `
"export type pet = {
    /**
     * @type integer, int64
    */
    id: number;
    /**
     * @type string
    */
    name: string;
    /**
     * @type string | undefined
    */
    tag?: string;
    category?: category;
};"
`;

exports[`typeGenerator schema > 'PetQuestionTokenAndUndefined' 1`] = `
"export type pet = {
    /**
     * @type integer, int64
    */
    id: number;
    /**
     * @type string
    */
    name: string;
    /**
     * @type string | undefined
    */
    tag?: string | undefined;
    category?: category | undefined;
};"
`;

exports[`typeGenerator schema > 'PetUndefined' 1`] = `
"export type pet = {
    /**
     * @type integer, int64
    */
    id: number;
    /**
     * @type string
    */
    name: string;
    /**
     * @type string | undefined
    */
    tag: string | undefined;
    category: category | undefined;
};"
`;

exports[`typeGenerator schema > 'Pets' 1`] = `
"export type pets = {
    /**
     * @type integer, int64
    */
    id: number;
    /**
     * @type string
    */
    name: string;
    /**
     * @type string | undefined
    */
    tag?: string;
}[];"
`;

exports[`typeGenerator schema > 'PetsStoreDiscriminator' 1`] = `
"/**
 * @description A project is an analysis project. It has a name, an id, and has a dataset
*/
export type petstore = {
    /**
     * @type integer
    */
    id: number;
    /**
     * @type array | undefined
    */
    pets?: (cat | dog)[];
};"
`;

exports[`typeGenerator schema > 'PetsStoreRef' 1`] = `"export type pets = pet[];"`;

exports[`typeGenerator schema > 'Plain_dateDate' 1`] = `"export type plainDate = Date;"`;

exports[`typeGenerator schema > 'Plain_dateString' 1`] = `"export type plainDate = string;"`;

exports[`typeGenerator schema > 'Plain_email' 1`] = `"export type plainEmail = string;"`;

exports[`typeGenerator schema > 'Plain_file' 1`] = `"export type plainFile = Blob;"`;

exports[`typeGenerator schema > 'Plain_timeDate' 1`] = `"export type plainTime = Date;"`;

exports[`typeGenerator schema > 'Plain_uuid' 1`] = `"export type plainUuid = string;"`;

exports[`typeGenerator schema > 'StringValueConst' 1`] = `
"/**
 * @description its value is equal to the value of the keyword
*/
export type stringValueConst = {
    /**
     * @type string | undefined
    */
    foobar?: "foobar";
};"
`;
